jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/internal/HttpStreamOverHTTP2.java

```diff
@@ -437,6 +437,20 @@ private void sendContent(MetaData.Request request, ByteBuffer content, boolean l
         }
     }
 
+    @Override
+    public Runnable cancelSend(Throwable cause, Callback appCallback)
+    {
+        // Append a ResetFrame into the H2 flusher and complete the appCallback once all the frames
+        // up to and including the reset one have been flushed; it is needed to wait to make sure
+        // the completion listeners aren't called while a write is still pending.
+        return () -> _stream.reset(new ResetFrame(_stream.getId(), ErrorCode.CANCEL_STREAM_ERROR.code), Callback.from(() ->
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("cancelSend reset callback complete");
+            appCallback.failed(cause);
+        }));
```

Comment:
`_stream.reset()` may still return early.

How about introducing a `HTTP2Session.flush(Callback)` that queues a "fake" `FlushEntry` into the `HTTP2FLusher` whose only scope is to be flushed and notify the callback when done?

Reply:
I'll do that, and I'll also write a test similar to `testResetAfterTCPCongestedWrite()` that waits for idle timeout instead of resolving the TCP congestion.

Reply:
Done.